pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2018-2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the adapter class for Qualtran-PennyLane interoperability.
+"""
+# pylint:disable=
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import MatrixUndefinedError, Operation
+from pennylane.wires import WiresLike
+
+try:
+    import qualtran as qt
+except (ModuleNotFoundError, ImportError) as import_error:
+    pass
+
+
+def get_bloq_registers_info(bloq):
+    """Returns a ``qml.registers`` object associated with all named and unnamed registers and wires
+    in the bloq.
+
+    Args:
+        bloq (Bloq): the bloq to get the registers info of
+
+    Returns:
+        dict: A dictionary that has all the named and un-named registers with default wire
+        ordering.
+
+    Raises:
+        TypeError: bloq must be an instance of :code:`~.Bloq`.
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    >>> from qualtran.bloqs.phase_estimation import RectangularWindowState, TextbookQPE
+    >>> textbook_qpe_small = TextbookQPE(ZPowGate(exponent=2 * 0.234), RectangularWindowState(3))
+    >>> qml.get_bloq_registers_info(textbook_qpe_small)
+    {'qpe_reg': Wires([0, 1, 2]), 'q': Wires([3])}
+    """
+    if not isinstance(bloq, qt.Bloq):
+        raise TypeError(f"bloq must be an instance of {qt.Bloq}.")
+
+    wire_register_dict = {reg.name: reg.bitsize for reg in bloq.signature.rights()}
+
+    return qml.registers(wire_register_dict)
+
+
+def _get_named_registers(registers):
+    """Returns a ``qml.registers`` object associated with the named registers in the bloq"""
+
+    temp_register_dict = {reg.name: reg.total_bits() for reg in registers}
+
+    return qml.registers(temp_register_dict)
+
+
+class FromBloq(Operation):
+    r"""
+    An adapter for using `Qualtran bloqs <https://qualtran.readthedocs.io/en/latest/bloqs/index.html#bloqs-library>`_
+    as a PennyLane :class:`~.Operation`.
+
+    Args:
+        bloq (qualtran.Bloq): the bloq to wrap
+        wires (WiresLike): the wires to act on
```

Comment:
I think these could have more detail. Below is an example of what I mean, but it should be adapted to your best judgement:
```
bloq (qualtran.Bloq): an initialized Qualtran bloq to simulate in a PennyLane
wires (WiresLike): the wires the operation acts on. This is determined by the Qualtran bloq signature.
```

Maybe include a minimal explanation of what "determined by the Qualtran signature" means.`CNOT` always requires two wires, but `Swap` depends on the `bitsize` argument fed to `Swap` on initialization. 

(By the way, the fact that `FromBloq` adapts based on the argument, and raises an informative error with the number of wires required is great. Joy to use :star: :cookie: )

Reply:
I can add a recommendation to the wires argument as well. Easiest way I've found to put in the # of wires is to just do Bloq.signature.n_qubits(). Maybe I'll mention that in the docstring as well.

Reply:
(Thanks Diego glad you liked it I made it myself)

Reply:
Could the user get this information from the helper function as well?

Reply:
Yeah they could. The helper function gives you the full registers (which is a dictionary), but since you can't pass in a dictionary, you would have to concatenate all the values in the dictionary to get the wires. The point of the helper function is to help users map qubits and that's what the helper function does.

I considered letting the function create both a dictionary and an array, but it ends up making the function confusing and overloaded in functionality. The alternative would be to create a separate helper function that essentially just returns `list(range(bloq.signature.n_qubits()))`, which idk if that's "useful" or necessary.